server/src/main/java/org/opensearch/cluster/coordination/JoinTaskExecutor.java

```diff
@@ -466,13 +477,30 @@ public static void ensureMajorVersionBarrier(Version joiningNodeVersion, Version
         }
     }
 
+    public static void ensureNodeNotDecommissioned(DiscoveryNode node, Metadata metadata) {
+        DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.custom(DecommissionAttributeMetadata.TYPE);
+        if (decommissionAttributeMetadata != null) {
+            DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+            if (decommissionAttribute != null) {
+                if (node.getAttributes().get(decommissionAttribute.attributeName()).equals(decommissionAttribute.attributeValue())) {
+                    throw new NodeDecommissionedException(
+                        "node has decommissioned attribute ["
+                            + decommissionAttribute.toString()
+                            + "]."
+                    );
+                }
+            }
+        }
+    }
+
     public static Collection<BiConsumer<DiscoveryNode, ClusterState>> addBuiltInJoinValidators(
         Collection<BiConsumer<DiscoveryNode, ClusterState>> onJoinValidators
     ) {
         final Collection<BiConsumer<DiscoveryNode, ClusterState>> validators = new ArrayList<>();
         validators.add((node, state) -> {
             ensureNodesCompatibility(node.getVersion(), state.getNodes());
             ensureIndexCompatibility(node.getVersion(), state.getMetadata());
+            ensureNodeNotDecommissioned(node, state.getMetadata());
```

Comment:
What does this achieve? Does it fail the discovery on the node?

Reply:
These are set of the join validators. Before any join request is executed, these set of validators validates the request on the leader node before processing the join.
The `DecommissionService` ensures that the decommissioned nodes are kicked out from the cluster. But as soon as a node leaves the cluster, it again starts the discovery and sends join request to the leader. This validator will prevent the decommissioned nodes to join back. The check is on the leader node and hence whenever the attribute is recommissioned, it will ensure the nodes join back in the next request